Null bytes in LIKE patterns are not escaped or rejected

## Description

Fuzzing discovered that null bytes (`\x00`) in LIKE patterns (used by `startsWith()` and `endsWith()`) are not properly handled. While we fixed null byte handling in string literals (#14), the LIKE pattern generation in string functions has the same vulnerability.

## Discovered By

- **CI Run**: https://github.com/SPANDigital/cel2sql/actions/runs/18400783446
- **Job ID**: 52429323134
- **Fuzzer**: `FuzzEscapeLikePattern`
- **Crash input**: `testdata/fuzz/FuzzEscapeLikePattern/390ca22614d4ce1a`

## Error

```
--- FAIL: FuzzEscapeLikePattern (0.31s)
    fuzz_test.go:251: Generated SQL contains null bytes: "name LIKE '\x00%'"
```

## Reproduction

```bash
go test -run=FuzzEscapeLikePattern/390ca22614d4ce1a
```

**Input pattern**: `"\x00"` (null byte)  
**CEL expression**: `name.startsWith("\x00")`  
**Generated SQL**: `name LIKE '\x00%'` (contains unescaped null byte)

## Impact

- **Security**: Null bytes in LIKE patterns can corrupt SQL queries
- **Correctness**: Pattern matching may behave unexpectedly
- **Consistency**: String literals reject null bytes (#14), but LIKE patterns don't

## Root Cause

The `callStartsWith()` and `callEndsWith()` functions in `cel2sql.go` process the pattern argument and generate LIKE clauses, but they don't check for null bytes before calling `escapeLikePattern()`.

The fix in #14 only addressed string literals in `visitConst()`, but didn't cover pattern arguments in string functions.

## Suggested Fix

Add null byte detection in `callStartsWith()` and `callEndsWith()` before generating LIKE patterns:

```go
func (con *converter) callStartsWith(target *exprpb.Expr, args []*exprpb.Expr) error {
    // ... existing code ...
    
    // Extract pattern string
    pattern := args[0].GetConstExpr().GetStringValue()
    
    // Reject patterns containing null bytes
    if strings.Contains(pattern, "\x00") {
        return errors.New("LIKE patterns cannot contain null bytes")
    }
    
    // ... rest of function ...
}
```

Apply the same fix to `callEndsWith()`.

## Alternative Solution

Use the same fix approach as #14 - rely on `visitConst()` to catch null bytes when processing the argument. However, this may not work if the pattern is computed or comes from a variable rather than a constant.

## Related Issues

- #12 - Original null byte bug in string literals (fixed in #14)
- PR #13 - Fuzzing infrastructure that discovered this issue
- PR #14 - Fixed null bytes in string literals, but missed LIKE patterns

## Test Coverage Needed

The existing `FuzzEscapeLikePattern` test already catches this issue. Once fixed, add explicit regression tests for:
- `name.startsWith("\x00")` → error
- `name.endsWith("\x00test")` → error
- `name.contains("test\x00")` → error

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

--- a/cel2sql.go
+++ b/cel2sql.go
@@ -279,6 +279,10 @@ func (con *converter) callStartsWith(target *exprpb.Expr, args []*exprpb.Expr) e
 	// If it's a constant string, we can append % directly
 	if constExpr := args[0].GetConstExpr(); constExpr != nil && constExpr.GetStringValue() != "" {
 		prefix := constExpr.GetStringValue()
+		// Reject patterns containing null bytes
+		if strings.Contains(prefix, "\x00") {
+			return errors.New("LIKE patterns cannot contain null bytes")
+		}
 		// Escape special LIKE characters: %, _, \
 		escaped := escapeLikePattern(prefix)
 		con.str.WriteString("'")
@@ -316,6 +320,10 @@ func (con *converter) callEndsWith(target *exprpb.Expr, args []*exprpb.Expr) err
 	// If it's a constant string, we can prepend % directly
 	if constExpr := args[0].GetConstExpr(); constExpr != nil && constExpr.GetStringValue() != "" {
 		suffix := constExpr.GetStringValue()
+		// Reject patterns containing null bytes
+		if strings.Contains(suffix, "\x00") {
+			return errors.New("LIKE patterns cannot contain null bytes")
+		}
 		// Escape special LIKE characters: %, _, \
 		escaped := escapeLikePattern(suffix)
 		con.str.WriteString("'%")

--- a/cel2sql_test.go
+++ b/cel2sql_test.go
@@ -562,3 +562,75 @@ func TestNullByteRejection(t *testing.T) {
 		})
 	}
 }
+
+// TestNullByteInLikePatterns tests that LIKE patterns with null bytes are rejected
+func TestNullByteInLikePatterns(t *testing.T) {
+	env, err := cel.NewEnv(
+		cel.Variable("name", cel.StringType),
+	)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name    string
+		expr    string
+		wantErr string
+	}{
+		{
+			name:    "startsWith with null byte at start",
+			expr:    `name.startsWith("\x00test")`,
+			wantErr: "LIKE patterns cannot contain null bytes",
+		},
+		{
+			name:    "startsWith with null byte in middle",
+			expr:    `name.startsWith("test\x00value")`,
+			wantErr: "LIKE patterns cannot contain null bytes",
+		},
+		{
+			name:    "startsWith with only null byte",
+			expr:    `name.startsWith("\x00")`,
+			wantErr: "LIKE patterns cannot contain null bytes",
+		},
+		{
+			name:    "endsWith with null byte at end",
+			expr:    `name.endsWith("test\x00")`,
+			wantErr: "LIKE patterns cannot contain null bytes",
+		},
+		{
+			name:    "endsWith with null byte in middle",
+			expr:    `name.endsWith("\x00value")`,
+			wantErr: "LIKE patterns cannot contain null bytes",
+		},
+		{
+			name:    "endsWith with only null byte",
+			expr:    `name.endsWith("\x00")`,
+			wantErr: "LIKE patterns cannot contain null bytes",
+		},
+		{
+			name:    "startsWith valid pattern",
+			expr:    `name.startsWith("valid")`,
+			wantErr: "",
+		},
+		{
+			name:    "endsWith valid pattern",
+			expr:    `name.endsWith("valid")`,
+			wantErr: "",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ast, issues := env.Compile(tt.expr)
+			require.NoError(t, issues.Err())
+
+			got, err := cel2sql.Convert(ast)
+			if tt.wantErr != "" {
+				require.Error(t, err)
+				assert.Contains(t, err.Error(), tt.wantErr)
+			} else {
+				require.NoError(t, err)
+				// Just verify it doesn't error - actual SQL format tested elsewhere
+				assert.NotEmpty(t, got)
+			}
+		})
+	}
+}

